src/managedwriter/stream_connection.ts

```diff
@@ -0,0 +1,268 @@
+      callOptions.otherArgs = {};
+    }
+    if (!callOptions.otherArgs.headers) {
+      callOptions.otherArgs.headers = {};
+    }
+    // This header is required so that the BigQuery Storage API
+    // knows which region to route the request to.
+    callOptions.otherArgs.headers[
+      'x-goog-request-params'
+    ] = `write_stream=${streamId}`;
+    return callOptions;
+  }
+
+  private handleData = (response: AppendRowsResponse) => {
+    const pw = this._pendingWrites.pop();
+    if (!pw) {
+      console.warn('data arrived with no pending write available', response);
+      return;
+    }
+    if (response.updatedSchema) {
+      this.emit('schemaUpdated', response.updatedSchema);
+    }
+    pw._markDone(null, response);
+  };
+
+  /**
+   * Callback is invoked when a the server notifies the stream connection
+   * of a new Table Schema change.
+   */
+  onSchemaUpdated(listener: (schema: TableSchema) => void): RemoveListener {
+    return this.registerListener('schemaUpdated', listener);
+  }
+
+  /**
+   * Callback is invoked when a write request fails based on an
+   * error received from the server.
+   */
+  onWriteError(
+    listener: (err: Error, req: AppendRowRequest) => void
+  ): RemoveListener {
+    return this.registerListener('writeError', listener);
+  }
+
+  /**
+   * Callback is invoked when an error is received from the server.
+   */
+  onConnectionError(listener: (err: Error) => void): RemoveListener {
+    return this.registerListener('error', listener);
+  }
+
+  private registerListener(
+    eventName: string,
+    listener: (...args: any[]) => void
+  ): RemoveListener {
+    this.addListener(eventName, listener);
+    return {
+      off: () => {
+        this.removeListener(eventName, listener);
+      },
+    };
+  }
+
+  // check if is default stream
+  private isDefaultStream(): boolean {
+    return this._streamId.endsWith('_default');
+  }
+
+  /**
+   * Get the name of the write stream associated with this connection.
+   * When the connection is created withouth a write stream,
+   * this method can be used to retrieve the automatically
+   * created write stream name.
+   */
+  getStreamId = (): string => {
+    return this._streamId;
+  };
+
+  /**
+   * Write a request to the bi-directional stream connection.
+   *
+   * @param {AppendRowRequest} request - request to send.
+   *
+   * @returns {managedwriter.PendingWrite}
+   */
+  write(request: AppendRowRequest): PendingWrite {
+    const pw = new PendingWrite(request);
+    if (!this._connection) {
+      pw._markDone(new Error('connection closed'));
+      return pw;
+    }
+    this._connection.write(request, err => {
+      if (err) {
+        this.emit('writeError', err, request);
+        pw._markDone(err); //TODO: add retries
+        return;
+      }
+      this._pendingWrites.unshift(pw);
+    });
+    return pw;
+  }
+
+  /**
+   * Check if connection is open and ready to send requests.
+   */
+  isOpen(): boolean {
```

Comment:
One thing I'm not seeing here (but I may be looking in the wrong place) is how the stream connection deals with connection closes not initiated by user behavior (ex: server closes connection after 10m idle).  How does recovery work here?

Reply:
Was added in the latest commit